firefoxThemes/resources.js

```diff
@@ -1,45 +1,209 @@
-          dokiTheme,
-        )
-      }), {})
-
-  }
+    constructor() {
+        this.initThemes();
+    }
+    /*Initialize Waifu Themes*/
+    initThemes() {
+        this.themes =
+            Object.entries(dokiThemeDefinitions).reduce((accum, [themeId, dokiTheme]) => ({
+                ...accum,
+                [themeId]: new Theme(
+                    dokiTheme.information.displayName,
+                    dokiTheme.information.imagePath,
+                    dokiTheme.information.jsonPath,
+                    dokiTheme,
+                )
+            }), {})
+    }
+}
+/*Location to retrieve resources for each waifu theme*/
+class Theme {
+    constructor(name, image, json, themeDefinition) {
+        this.name = name;//Name of theme
+        this.image = image;//Relative link to waifu image file
+        this.json = json;//Relative link to browser theme file
+        this.definition = themeDefinition;
+        this.page = "/waifus/index.html";
+    }
+}
+/*---FUNCTIONS---*/
+/*Initialize Local Storage & custom new tab page*/
+async function startStorage() {
+    browser.storage.local.get("currentThemeId")
+        .then((storage) => {
+            const initStorage = {
+                waifuThemes: new WaifuThemes()
+            };
+            //Retrieve all themes if none exists in local storage
+            browser.storage.local.set(initStorage);
+            //Load browser theme
+            if (storage.currentThemeId) {
+                loadTheme(initStorage.waifuThemes.themes, storage.currentThemeId);
+            }
+            //When the browser first opens, redirect to custom new tab page
+            browser.tabs.update({loadReplace: true, url: "waifus/index.html"});
+        });
+}
+/*Update all new tabs with new waifu theme*/
+function updateTabs(msg) {
+    switch (msg.mixState){
+        case mixedStates.INITIAL:
+            SetupMixedUpdate();
+            break;
+        case mixedStates.TAB_CREATED:
+            MixedUpdate(msg);
+            break;
+        default:
+            MixTabCleanup();
+            NormalUpdate(msg);
+            break;
+    }
+}
+/*Initialize the Mixed feature*/
+function SetupMixedUpdate(){
+    browser.tabs.query({title:"New Tab"})
+        .then((tabs)=>{
+            browser.storage.local.get("waifuThemes")
+                .then((storage)=>{
+                    if(!mixedTabs){
+                        mixedTabs = new Map();//Create a new mixed tab list
+                    }
+                    //Activate event listeners
+                    browser.tabs.onCreated.addListener(MixTabCreated);//When a tab is created
+                    browser.tabs.onActivated.addListener(MixTabActivated);//When the active tab has been switched
+                    browser.tabs.onRemoved.addListener(MixTabClosed);//When a tab has been closed
+                    const themes = storage.waifuThemes.themes;
+                    let currentThemeId = getRandomTheme(themes);
```

Comment:
Was there a reason why each existing tab didn't get a different theme? (Did it have something to do with the background not changing?)

Reply:
Firefox does not have a built in function where each tab can have its own separate theme. There can only be one browser theme in use at a time.

Reply:
So a workaround was made

Reply:
What I was trying to ask was why not assign pre-existing tabs a new random theme, so that way when the user navigates to it changes the theme. Right now when I set mixed mode it just applies the same random theme.

![Peek 2020-12-30 16-15](https://user-images.githubusercontent.com/15972415/103383960-43fda180-4aba-11eb-9dcc-8a07b92fcdf5.gif)

Reply:
I see. Did not know this occurs. Thanks for pointing it out.